### PR TITLE
feat: reward lock processing

### PR DESCRIPTION
**Pull request by r4mmer**
### Acceptance Criteria
- We should have a separate list of locked utxos to avoid iterating on all utxos.
- When the height is updated we should process the locked utxos
- Create a method to process locked utxos that became unlocked

### Security Checklist
- [ ] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

**Comment by r4mmer** [comment]
closes #478

File: src/utils/connection.ts [hunk 1]
```diff
@@ -10,7 +10,9 @@ import { IStorage } from "../types";
 export function handleWsDashboard(storage: IStorage) {
   return (data: any) => {
     // update network height
-    storage.setCurrentHeight(data.best_block_height as number);
+    const height = data.best_block_height as number;
+    storage.setCurrentHeight(height);
+    storage.unlockUtxos(height);
```
Comment (pedroferreira1) [inline src/utils/connection.ts:15]: This ws message is sent every second. I think we should check if the height has changed before doing anything here, what do you think?

File: src/storage/storage.ts [hunk 2]
```diff
@@ -46,12 +46,22 @@ export class Storage implements IStorage {
...
   config: Config;
   version: ApiVersion|null;
+  /**
+   * This promise is used to chain the calls to process unlocked utxos.
+   * This way we can avoid concurrent calls.
+   * The best way to do this would be an async queue or a mutex, but to avoid adding
+   * more dependencies we are using this simpler method.
+   *
+   * We can change this implementation to use a mutex or async queue in the future.
+   */
```
Comment (pedroferreira1) [inline src/storage/storage.ts:56]: Nice comment 👍

File: src/storage/storage.ts [hunk 3]
```diff
@@ -249,6 +259,21 @@ export class Storage implements IStorage {
     await this.store.unregisterToken(tokenUid);
   }
 
+  /**
+   * Process the locked utxos to unlock them if the lock has expired.
+   * Will process both timelocked and heightlocked utxos.
+   *
+   * We will wait for the any previous execution to finish before starting the next one.
```
Comment (pedroferreira1) [inline src/storage/storage.ts:266]:
```suggestion
   * We will wait for any previous execution to finish before starting the next one.
```

File: __tests__/storage/storage.test.js [hunk 4]
```diff
@@ -143,3 +143,230 @@ test('utxos selected as inputs', async () => {
...
+      address,
+      value: output.value,
+      authorities: 0,
+      timelock,
+      type: tx.version,
+      height: tx.height,
+    };
+  }
+
+  const store = new MemoryStore();
```
Comment (pedroferreira1) [inline __tests__/storage/storage.test.js:198]: Can we do the same test for leveldb store?

File: __tests__/storage/memory_store.test.js [hunk 5]
```diff
@@ -262,6 +262,125 @@ test('utxo methods', async () => {
...
+      type: tx.version,
+      height: tx.height,
+    };
+  }
+
+  jest
+   .useFakeTimers()
+   .setSystemTime(tsCurrent);
+
+  const store = new MemoryStore();
```
Comment (pedroferreira1) [inline __tests__/storage/memory_store.test.js:320]: Can you do the same tests for the leveldb store?

File: src/utils/connection.ts (hunk 1 above)
Comment (r4mmer) [reply src/utils/connection.ts]: Started to check if the current height is different than the new height before calling `unlockUtxos`

File: src/storage/storage.ts (hunk 3 above)
Comment (r4mmer) [reply src/storage/storage.ts]: Done

File: __tests__/storage/storage.test.js (hunk 4 above)
Comment (r4mmer) [reply __tests__/storage/storage.test.js]: Done

File: __tests__/storage/memory_store.test.js (hunk 5 above)
Comment (r4mmer) [reply __tests__/storage/memory_store.test.js]: Done.

File: __tests__/storage/common_store.test.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { LevelDBStore, MemoryStore } from "../../src/storage";
+import { HDPrivateKey } from "bitcore-lib";
+import { TOKEN_AUTHORITY_MASK, TOKEN_MINT_MASK } from "../../src/constants";
+
+const DATA_DIR = './testdata.leveldb';
+
+describe("locked utxo methods", () => {
```
Comment (r4mmer) [inline __tests__/storage/common_store.test.js:14]: I will start to port the other store tests to this file with a similar approach in another PR.

Basically we should have a common tests that can run with any store implementation to make sure all `IStore` implementations behave the same way.
Ideally we should do the same with `IStorage` implementations and check that any combination of `IStorage` and `IStore` will work as intended.

**Review by tuliomir — APPROVED** [review]

File: __tests__/storage/common_store.test.js (hunk 6 above)
Comment (tuliomir) [reply __tests__/storage/common_store.test.js:14]: I agree. The test codes got too similar as well, it will be good to have them unified.

Comment (r4mmer) [reply __tests__/storage/common_store.test.js:14]: https://github.com/HathorNetwork/hathor-wallet-lib/issues/488

File: src/utils/connection.ts [hunk 7]
```diff
@@ -11,8 +11,12 @@ export function handleWsDashboard(storage: IStorage) {
   return (data: any) => {
     // update network height
     const height = data.best_block_height as number;
-    storage.setCurrentHeight(height);
-    storage.unlockUtxos(height);
+    storage.getCurrentHeight().then(async (currentHeight) => {
+      await storage.setCurrentHeight(height);
```
Comment (pedroferreira1) [inline src/utils/connection.ts:15]: You could move this inside the `if` below as well but it's fine anyway.

File: src/utils/connection.ts [hunk 8]
```diff
@@ -11,8 +11,12 @@ export function handleWsDashboard(storage: IStorage) {
   return (data: any) => {
     // update network height
     const height = data.best_block_height as number;
-    storage.setCurrentHeight(height);
-    storage.unlockUtxos(height);
+    storage.getCurrentHeight().then(async (currentHeight) => {
```
Comment (pedroferreira1) [inline src/utils/connection.ts:14]: Is there a reason to use promise here?

File: src/utils/connection.ts [hunk 9]
```diff
@@ -11,8 +11,12 @@ export function handleWsDashboard(storage: IStorage) {
...
+      await storage.setCurrentHeight(height);
+      if (height !== currentHeight) {
+        await storage.unlockUtxos(height);
```
Comment (pedroferreira1) [inline src/utils/connection.ts:17]: Do we need to await?

**Review by pedroferreira1 — APPROVED** [review]

File: src/utils/connection.ts (hunk 8 above)
Comment (r4mmer) [reply src/utils/connection.ts]: Its only needed to await the promises below, but they are not actually necessary.

File: src/utils/connection.ts (hunk 9 above)
Comment (r4mmer) [reply src/utils/connection.ts]: Not really, the initial idea was to run `unlockUtxos` before changing the height, but passing the height as a parameter is better, but it made the `await`s unnecessary, so I'm removing them

**Review by pedroferreira1 — APPROVED** [review]

**Pull request merged**